[PyTorch] Add Expanded call stack to nodes

RihamSelim opened the pull request:
Summary:
To get a Node's call stack we currently loop on the InlinedCallStack graph and follow the "callee" chain. Since the node's inlined stack does not change we can optimize this but expanding the node's inlined stack once and reusing it. This is particularly useful when reading the node's stack from another process (e.g. BPF) as it simplified the memory traversal process.

The new data structure (NodeSourceInfo) only holds pointers to the function name and file name variables, and assumes these objects will be alive throughout the lifetime of the process.

Each Node has an extended attribute that has an index to a vector of stack frames `expanded_node_stacks_`

`node_stack_attr_symbol_` is only needed to make accessing the stack vector index attribute easier from BPF.

Test Plan:
- Performance Impact: The cost of expanding the call stack is between 500 - 1000 ns and happens only per instruction node at initialization time.
- Verified using BPF Program in subsequent diffs

Reviewed By: zdevito

Differential Revision: D46578700

RihamSelim commented:
/easycla

RihamSelim commented:
/easycla

zdevito commented on torch/csrc/jit/runtime/interpreter/code_impl.h:
```diff
@@ -213,13 +224,64 @@ struct CodeImpl {
+
+    SourceRange r = range;
+    if (range.source()) {
+      if (auto orig = range.source()->findSourceRangeThatGenerated(r)) {
```
I am wondering about the cost of `findSourceRangeThatGenerated` when called on all lines of code. Can this be run on a model that includes  original source ranges and measured the total loading time with/without this preprocessing step? The code is going to read and unpickle some debug files that would not otherwise have been done, which I think can significantly increase the loading time if those files are big. I didn’t notice before that this pathway is only done lazily but includes file reading.

It would be fine to land this PR without this line, and just record the stack traces without original source code. It would also be fine to measure the overall cost for some big model with source ranges and figure out that only adds a small amount of time to decoding.

RihamSelim replied:
PR updated to split `findSourceRangeThatGenerated` into a separate PR

Skylion007 commented on torch/csrc/jit/runtime/interpreter/code_impl.h:
```diff
@@ -59,15 +59,26 @@ struct WithCurrentNode {
 
+
+struct NodeSourceInfo {
+  const char* func_name_;
```
You should use use nullptr instead of NULL + optional<std::string> for less chance accidental bugs / mem leaks.

Skylion007 reviewed (commented):
C++ nits

RihamSelim committed:
[PyTorch] Add Expanded call stack to nodes (#108426)

Summary:

To get a Node's call stack we currently loop on the InlinedCallStack graph and follow the "callee" chain. Since the node's inlined stack does not change we can optimize this but expanding the node's inlined stack once and reusing it. This is particularly useful when reading the node's stack from another process (e.g. BPF) as it simplified the memory traversal process.

The new data structure (NodeSourceInfo) only holds pointers to the function name and file name variables, and assumes these objects will be alive throughout the lifetime of the process.

Each Node has an extended attribute that has an index to a vector of stack frames `expanded_node_stacks_`

`node_stack_attr_symbol_` is only needed to make accessing the stack vector index attribute easier from BPF.

Test Plan:
- Performance Impact: The cost of expanding the call stack is between 500 - 1000 ns and happens only per instruction node at initialization time.
- Verified using BPF Program in subsequent diffs

Reviewed By: zdevito

Differential Revision: D46578700

zdevito commented:
@pytorchbot merge

zdevito commented:
@pytorchbot merge -f "unrelated failure"

clee2000 commented:
@pytorchbot revert -m "got reverted internally" -c weird
